Add basic tracking of the changes in a tree update

mwcampbell opened the pull request:
The code for processing a `TreeUpdate` can now keep track of the nodes that are added, updated, or removed, as well as a change in focus. This feature will be the basis for platform-specific events.

DataTriny reviewed (commented):
Overall you are on the right track, just some comments on naming and tests.

mwcampbell commented:
I changed the names of the variants, and switched to named fields where appropriate.

I agree that the tests are awkward. But in the current design, I can't collect the changes in a `Vec` because some of the variants implicitly borrow a `tree::Reader` which is only alive during the `update_and_process_changes` function. That's why I decided to use a callback in the first place. I'm open to something better than `assert!(false)` though.

DataTriny commented on consumer/src/tree.rs:
```diff
@@ -301,7 +392,28 @@ mod tests {
+                    return;
+                }
+            }
+            assert!(false);
```
You could directly `panic!` with a message such as: Expected root node update and new child changes. At least it will be clear from the test results if it ever break.

DataTriny commented on consumer/src/tree.rs:
```diff
@@ -334,7 +446,28 @@ mod tests {
+                    return;
+                }
+            }
+            assert!(false);
```
`panic!("Expected root node update and child removal changes.");`

DataTriny commented on consumer/src/tree.rs:
```diff
@@ -369,8 +502,99 @@ mod tests {
+                    }
+                }
+            }
+            assert!(false);
```
`panic!("Expected focus moved change.");`
